File: Computer networks/socket/thread/function_thread.c
// Nome Cognome  - data - versione
// Il programma e' suddiviso in diverse unita' funzionali gestite da diversi thread.
// In questo esempio abbiamo un processo  interattivo e 2  unita' f1, f2.
// gcc function_thread.c -o function_thread -pthread  
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>    // usleep
#include <sys/time.h>  // gettimeofday
#include <pthread.h>

void * f1(void *p);     // Thread f1
void * f2(void *p);     // Thread f2

pthread_t mythread1;
pthread_t mythread2;

/////////////////////// main() //////////////////////////

int main(int argc, char **argv) {
   int S1 = 12;
   int S2 = 20;

   pthread_create(&mythread1, NULL, f1, (void *)&S1); 
   pthread_create(&mythread2, NULL, f2, (void *)&S2); 

   char answ[20]; 

   do {
      printf("Prompt> ");
      fgets(answ,20,stdin);
   } while (answ[0]!='Q' && answ[0]!='q');

   printf("Wait for threads..\n");
   pthread_join(mythread1, NULL); 
   pthread_join(mythread2, NULL); 
   printf("Threads finished. Exiting\n");

   return (0);
}

//////////////////////////////////////////////////////////////

void * f1(void *p) {
   int n = *(int *)p; 
   printf ("F1  start. Going to sleep for %d seconds \n", n);
   sleep(n); 
   printf ("F1 stop. \n", n);
   return (0); 
}

//////////////////////////////////////////////////////////////

void * f2(void *p) {
   int n = *(int *)p;
   printf ("F2  start. Going to sleep for %d seconds \n", n);
   sleep(n);
   printf ("F2 stop. \n", n);
   return (0);
}

File: Computer networks/socket/thread/function_thread.py
#!/usr/bin/env python

# function_thread.py
#
# Ayoub Ouarrak
# 17-11-2014
# Il programma e' suddiviso in diverse unita' funzionali gestite da diversi thread.
# In questo esempio abbiamo un processo  interattivo e 2  unita' f1, f2.

from threading import Thread
from time import sleep
import sys;

def thread_f1(arg):
    print "Thread f1 started. Sleep for  " + str(arg) + " seconds"
    sleep(arg)
    print "Thread f1 finished"

def thread_f2(arg):
    print "Thread f2 started. Sleep for  " + str(arg) + " seconds"
    sleep(arg)  
    print "Thread f2 finished"                

def interactive():
  print "Interacive started (Press q to exit)"
  while True:
    cmd = raw_input(">> ")
    if cmd.strip() == '':
        continue
    if cmd.strip() == 'Q' or cmd.strip() == 'q':
        break
    print "command: " + cmd 


if __name__ == "__main__":
    t1 = 20
    t2 = 12

    if len(sys.argv) == 2:
        t1 = int(sys.argv[1])

    if len(sys.argv) == 3:
        t1 = int(sys.argv[1])
        t2 = int(sys.argv[2])

    if len(sys.argv) > 3:
        print "Usage: "
        print "python function_thread.py [t1 t2]"
        print "defaults value of t1 = 20, t2 = 12"
        SystemExit("too many arguments")

    thread1 = Thread(target = thread_f1, args = (t1, ))
    thread2 = Thread(target = thread_f2, args = (t2, ))

    thread1.start()
    thread2.start()

interactive()

# wait until the others thread has completed its work
thread1.join()
thread2.join()
print "Functional multithread finished, exiting...."

